core/trino-main/src/main/java/io/trino/operator/exchange/UniformPartitionRebalancer.java

```diff
@@ -0,0 +1,430 @@
+        }
+    }
+
+    private class RebalanceContext
+    {
+        private final Set<Integer> rebalancedPartitions = new HashSet<>();
+        private final long[] writerPhysicalWrittenBytesSinceLastRebalance;
+        private final long[] writerRowCountSinceLastRebalance;
+        private final long[] writerPriorities;
+        private final List<IndexedPriorityQueue<PartitionIdWithRowCount>> writerMaxPartitions;
+
+        private RebalanceContext(List<Long> writerPhysicalWrittenBytes, Long2LongMap partitionRowCounts)
+        {
+            writerPhysicalWrittenBytesSinceLastRebalance = new long[numberOfWriters];
+            writerPriorities = new long[numberOfWriters];
+            for (int writerId = 0; writerId < writerPhysicalWrittenBytes.size(); writerId++) {
+                long physicalWrittenBytesSinceLastRebalance =
+                        writerPhysicalWrittenBytes.get(writerId) - writerPhysicalWrittenBytesAtLastRebalance.get(writerId);
+                writerPhysicalWrittenBytesSinceLastRebalance[writerId] = physicalWrittenBytesSinceLastRebalance;
+                writerPriorities[writerId] = physicalWrittenBytesSinceLastRebalance;
+            }
+
+            writerRowCountSinceLastRebalance = new long[numberOfWriters];
+            writerMaxPartitions = new ArrayList<>(numberOfWriters);
+            for (int writerId = 0; writerId < numberOfWriters; writerId++) {
+                writerMaxPartitions.add(new IndexedPriorityQueue<>());
+            }
+
+            partitionRowCounts.forEach((serializedKey, rowCount) -> {
+                WriterPartitionId writerPartitionId = WriterPartitionId.deserialize(serializedKey);
+                writerRowCountSinceLastRebalance[writerPartitionId.writerId] += rowCount;
+                writerMaxPartitions
+                        .get(writerPartitionId.writerId)
+                        .addOrUpdate(new PartitionIdWithRowCount(writerPartitionId.partitionId, rowCount), rowCount);
+            });
+        }
+
+        private boolean rebalancePartition(WriterId from, WriterId to)
+        {
+            IndexedPriorityQueue<PartitionIdWithRowCount> maxPartitions = writerMaxPartitions.get(from.id);
+
+            while (true) {
+                // Find the partition with maximum written bytes since last rebalance
+                PartitionIdWithRowCount partitionToRebalance = maxPartitions.poll();
+
+                if (partitionToRebalance == null) {
+                    return false;
+                }
+
+                PartitionInfo partitionInfo = partitionInfos[partitionToRebalance.id];
+
+                // If partition is already balanced or min skewed writer is already writing to that partition, then skip
+                // this partition and move on to the other partition inside max writer.
+                if (isPartitionRebalanced(partitionToRebalance.id) || partitionInfo.containsWriter(to.id)) {
```

Comment:
This is potentially slow if we consider default `minWriterSize` is 32MB while rebalance threshold is 100MB (e.g. when writing to a single partition)